cpp/dolfinx/geometry/BoundingBoxTree.cpp

```diff
@@ -20,381 +20,331 @@ using namespace dolfinx::geometry;
-    const double* pi = points[i].data();
-    const double* pj = points[j].data();
-    return pi[axis] < pj[axis];
-  });
+  std::array<double, 3> b_diff;
+  std::transform(b1.begin(), b1.end(), b0.begin(), b_diff.begin(),
+                 std::minus<double>());
+  const std::size_t axis = std::distance(
+      b_diff.begin(), std::max_element(b_diff.begin(), b_diff.end()));
+  auto middle = std::next(points.begin(), points.size() / 2);
+  std::nth_element(
+      points.begin(), middle, points.end(),
+      [axis](const std::pair<std::array<double, 3>, std::int32_t>& p0,
+             const std::pair<std::array<double, 3>, std::int32_t>& p1) -> bool {
+        return p0.first[axis] < p1.first[axis];
+      });
 
   // Split bounding boxes into two groups and call recursively
-  std::array bbox{
-      _build_from_point(points, begin, middle, bboxes, bbox_coordinates),
-      _build_from_point(points, middle, end, bboxes, bbox_coordinates)};
+  std::array bbox{_build_from_point(tcb::span(points.begin(), middle), bboxes,
+                                    bbox_coordinates),
+                  _build_from_point(tcb::span(middle, points.end()), bboxes,
+                                    bbox_coordinates)};
 
-  // Store bounding box data. Note that root box will be added last
+  // Store bounding box data. Note that root box will be added last.
   bboxes.push_back(bbox);
-  bbox_coordinates.insert(bbox_coordinates.end(), b.data(), b.data() + 3);
-  bbox_coordinates.insert(bbox_coordinates.end(), b.data() + 3, b.data() + 6);
+  bbox_coordinates.insert(bbox_coordinates.end(), b0.begin(), b0.end());
+  bbox_coordinates.insert(bbox_coordinates.end(), b1.begin(), b1.end());
   return bboxes.size() - 1;
 }
 //-----------------------------------------------------------------------------
 } // namespace
 
 //-----------------------------------------------------------------------------
-BoundingBoxTree::BoundingBoxTree(
-    const Eigen::Array<int, Eigen::Dynamic, 2, Eigen::RowMajor>& bboxes,
-    const Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor>& bbox_coords)
-    : _tdim(0), _bboxes(bboxes), _bbox_coordinates(bbox_coords)
+BoundingBoxTree::BoundingBoxTree(const mesh::Mesh& mesh, int tdim,
+                                 double padding)
+    : BoundingBoxTree::BoundingBoxTree(mesh, tdim, range(mesh, tdim), padding)
 {
   // Do nothing
 }
 //-----------------------------------------------------------------------------
 BoundingBoxTree::BoundingBoxTree(const mesh::Mesh& mesh, int tdim,
+                                 const tcb::span<const std::int32_t>& entities,
                                  double padding)
     : _tdim(tdim)
 {
-  // Check dimension
-  if (tdim > mesh.topology().dim())
+  if (tdim < 0 or tdim > mesh.topology().dim())
   {
-    throw std::runtime_error("Dimension must be a number between 0 and "
+    throw std::runtime_error("Dimension must be a number between 1 and "
```

Comment:
This should still be zero? It is allowed to build a vertex-tree? At least the if test does not fail if tdim==0.

Reply:
Tests fail if `tdim==0` is not allow, e.g. the test at https://github.com/FEniCS/dolfinx/blob/master/python/test/unit/geometry/test_bounding_box_tree.py#L177 fails.

Reply:
So what I meant is that the runtime error should not say that dimensions is a number between 1 and tdim, but 0 and tdim